### ISSUE TITLE
Inconsistent coloring between logback and log4j

### ISSUE THREAD

**Issue by kilemensi** [jcgay/maven-color#9]
![log4j2](https://cloud.githubusercontent.com/assets/1779590/7440849/1d87b5f8-f0d5-11e4-9940-a58cd815b91d.png)
![logback](https://cloud.githubusercontent.com/assets/1779590/7440850/1e2f3dc8-f0d5-11e4-9398-48d32cfecc71.png)

The two implementations give different color outputs (I believe log4j2 is the correct one but I hate the little warning it gives about not being supported by maven)

Cheers.

**Comment by jcgay** [comment]
Thanks for reporting this !

The coloring is really strange yes, with the two different implementations by the way.
I wasn't aware of http://mridgers.github.io/clink/, I will try it to reproduce the issue.

Do you have the same behavior with a classic `cmd` ?

For the Log4J 2 implementation, I think I may get rid of it, I will try to use newer version but I remember failing last time I did.

**Comment by kilemensi** [comment]
For classic cmd
 + `Logback` doesn't give any colors, just ANSI codes I think
 + `Log4J 2` works as expected (just the warning)


![logback](https://cloud.githubusercontent.com/assets/1779590/7441470/afec19cc-f0f1-11e4-9d8b-a9fb0ec48078.png)

![log4j2](https://cloud.githubusercontent.com/assets/1779590/7441471/b182f648-f0f1-11e4-8e40-a5c78c767f1b.png)

**Comment by jcgay** [comment]
In your classic `cmd` (How do you launch it by the way after `Clink` installation ?) can you give me the result of `echo %TERM%` ?

Here what I got with `logback`, classic `cmd`:
![maven-color-cmd](https://cloud.githubusercontent.com/assets/967336/7445539/9fa70320-f1b8-11e4-9e43-2d6814e50e9d.png)

and `Clink`:
![maven-color-clink](https://cloud.githubusercontent.com/assets/967336/7445540/aa09fb4c-f1b8-11e4-86f6-725965f7aac9.png)

But in you first screens, it seems that you are using other fancy window tools :) ? What are they ? You have some kind of theme which is changing console color no ?

**Comment by kilemensi** [comment]
`echo %TERM%` give `msys`

In first screens, I was using [ConEmu](http://conemu.github.io/).
In the second set of screens, I was running raw cmd with out anything (not even `Clink`).

**Comment by jcgay** [comment]
Ok thanks.

For your terminal displaying ANSI escape character with `logback`, I think it comes from #6.
I have assumed that if you terminal has env variable `%TERM%` set, it will support colors.
But that's a relative naive shortcut :disappointed:

Can you verify that modifying:

    <withJansi>${withJansi}</withJansi>

with

    <withJansi>true</withJansi>

in  `$M2_HOME/conf/logging/logback.xml` solves the problem ?

If yes I will just start maintaining a list of Windows terminal supporting colors without `Jansi` :stuck_out_tongue: 

I will give a try to ConEmu !

**Comment by kilemensi** [comment]
Yes, setting it manually does the trick. See below pick for both raw CMD and ConEmu (with Clink)

![CMD](https://cloud.githubusercontent.com/assets/1779590/7446930/5f6b949a-f1f1-11e4-9014-a7f488931e50.png)
![ConEMu with Clink](https://cloud.githubusercontent.com/assets/1779590/7446946/b9b5970c-f1f1-11e4-819c-4cc1050be959.png)

**Comment by jcgay** [comment]
Thanks !

I have built a new version, can you try it in your environment ?
https://dl.dropboxusercontent.com/u/2247464/maven-color-logback-1.1-SNAPSHOT-bundle.zip
https://dl.dropboxusercontent.com/u/2247464/maven-color-log4j2-1.1-SNAPSHOT-bundle.zip

**Comment by kilemensi** [comment]
The new build does solve the problem and now the two versions give out consistent coloring (see below pics)
*`Log4j2`*
![log4j2](https://cloud.githubusercontent.com/assets/1779590/7453072/278d80d8-f26e-11e4-97a3-73b17e8c01a0.png)

*`Logback`*
![logback](https://cloud.githubusercontent.com/assets/1779590/7453110/7a414efe-f26e-11e4-91fb-74ead59a9a19.png)

But since I'm a NetBeans user, which practically runs on maven, testing with the two frameworks works perfectly for `log4j2` but `logback` given an error/warning about failing to create `WindowsAnsiOutputStream` but continues to work afterwards on NetBeans Output window. Full ouput:
```
15:04:29,169 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - Failed to create WindowsAnsiOutputStream. Falling back on the default stream. ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.fusesource.jansi.WindowsAnsiOutputStream
	at ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.fusesource.jansi.WindowsAnsiOutputStream
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassNameAndParameter(OptionHelper.java:73)
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassNameAndParameter(OptionHelper.java:42)
	at 	at ch.qos.logback.core.ConsoleAppender.getTargetStreamForWindows(ConsoleAppender.java:94)
	at 	at ch.qos.logback.core.ConsoleAppender.start(ConsoleAppender.java:85)
	at 	at ch.qos.logback.core.joran.action.AppenderAction.end(AppenderAction.java:96)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.callEndAction(Interpreter.java:317)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.endElement(Interpreter.java:196)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.endElement(Interpreter.java:182)
	at 	at ch.qos.logback.core.joran.spi.EventPlayer.play(EventPlayer.java:62)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:149)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:135)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:99)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:49)
	at 	at ch.qos.logback.classic.util.ContextInitializer.configureByResource(ContextInitializer.java:77)
	at 	at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:152)
	at 	at org.slf4j.impl.StaticLoggerBinder.init(StaticLoggerBinder.java:85)
	at 	at org.slf4j.impl.StaticLoggerBinder.<clinit>(StaticLoggerBinder.java:55)
	at 	at org.slf4j.LoggerFactory.bind(LoggerFactory.java:128)
	at 	at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:107)
	at 	at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:295)
	at 	at org.apache.maven.cli.MavenCli.logging(MavenCli.java:431)
	at 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:276)
	at 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
	at 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at 	at java.lang.reflect.Method.invoke(Method.java:497)
	at 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: java.lang.reflect.InvocationTargetException
	at 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at 	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassNameAndParameter(OptionHelper.java:68)
	at 	... 30 common frames omitted
Caused by: java.io.IOException: Could not get the screen info: T h e   o p e r a t i o n   c o m p l e t e d   s u c c e s s f u l l y .

	at 	at org.fusesource.jansi.WindowsAnsiOutputStream.getConsoleInfo(WindowsAnsiOutputStream.java:104)
	at 	at org.fusesource.jansi.WindowsAnsiOutputStream.<init>(WindowsAnsiOutputStream.java:97)
	at 	... 35 common frames
```

**Comment by jcgay** [comment]
Thanks for the time you spent testing all this :+1: 

I have found an issue opened [here](http://jira.qos.ch/browse/LOGBACK-762) with the same kind of warning you are facing.
You still have colors in Netbeans even with the warning right ? Maybe you can configure Logback to ignore this logger ?
If not I may have a look to write a custom `ConsoleAppender` as suggested in the issue.

**Comment by kilemensi** [comment]
Nah, it's no problem. I enjoy using it.


Yes, you still get colors after the warning.


I've tried disabling/ignoring the logger using `<logger ... level="off/error">` to no effect since logback prints the warning output *before* setting the logger level on this specific logger.


But when testing this I noticed something else that I didn't notice before since I was only testing the output by compiling. This time I did a full clean/install. I don't know if you use NetBeans or not but i a nutshell when you run tests, when the actual test run after compiling is complete, NetBeans switches from *Output* window into a more graphical *Test Results* windows. When running normal maven (no ANSI color codes) this happens fine but using `maven-color` (with either *Logback* or *Log4j2*) this fails. NetBeans remains on the *Output* window and test results are just shown textually and not graphically.

The only thing I could think that could cause this is if NetBeans parses the output of maven and uses that to trigger the switch and hence presence of ansi codes somehow interferes with this. The only test I could do is create a custom `ConsoleAppender` as in the issue you linked and tested with that. And it actually works. Since the custom `ConsoleAppender` does not insert any ansi code, coloring disappears as a result. But this is not a big issue since NetBeans allows to customize it's output window with different colors (error, debug, etc.) and hence I don't think `maven-color` should bother with it. The only remaining issue is the warning message still get printed (I think it happens when logback is actually bootstraping or something so it's difficult to suppress).

**Comment by jcgay** [comment]
Ok I think this is time I install Netbeans in my VM :smile: 

Thanks again for your analyze !

**Comment by jcgay** [comment]
I think I've come with an implementation of `ConsoleAppender` for Logback that hide the warning.
From my testing with Netbeans, I haven't got any warning and the test report is automatically opened when test run is over.

I have uploaded a new build at: https://dl.dropboxusercontent.com/u/2247464/maven-color-logback-1.1-SNAPSHOT-bundle.zip
If you can give it a try, that would be awesome ! Thanks !

**Comment by jcgay** [comment]
I will test Log4J 2 soon...

**Comment by kilemensi** [comment]
Tested it and it works perfectly... Good work.

I'll be using this logback snapshot version until the Log4J 2 release. Cheers.

**Comment by jcgay** [comment]
I have tested with Log4J 2 and I didn't get the previous stacktrace we had with Logback.
But I have the incorrect behavior meaning the test output in Netbeans is not opened.

I have tried adding `-Dlog4j.skipJansi=true` in the Global execution options field in Maven Netbeans configuration screen, but it doesn't seems to work... https://logging.apache.org/log4j/2.x/manual/configuration.html#System_Properties

Extending `ConsoleAppender` is not possible with Log4J.

I'm going to release maven-color this way. If you have a better idea feel free to open a new issue to make it works with Netbeans.

Thanks again !
